### PR TITLE
inv: add ability to restart level from passport when Lara dies

### LINKED ISSUE #48: Feature request: restart level feature
Relevant TR2M commit: https://github.com/Arsunt/TR2Main/commit/0527c41bf0e2fa17e3d45232ff7c44561426d8aa

### PR DESCRIPTION
**Pull request by walkawayy**
Closes #48.

### CODE REVIEW

**Comment by walkawayy** [comment]
The one thing I don't fully understand is when the game restarts from `StopGame` in `game.c` vs. `Display_Inventory` in `inventry.c`. But this code works for all the scenarios I could think of.

**Review by rr- — COMMENTED** [review]
What happens to
- the pickups Lara collected
- the guns Lara found
- the secrets Lara found
- the ammo Lara wasted
- the medpacks Lara consumed
- the health Lara lost or gained
- 
between starting a particular level and dying and restarting it?

How does it play with the "keep Lara's health between level restarts" feature?

**Comment by walkawayy** [comment]
> 
> 
> What happens to
> 
>     * the pickups Lara collected
> 
>     * the guns Lara found
> 
>     * the secrets Lara found
> 
>     * the ammo Lara wasted
> 
>     * the medpacks Lara consumed
> 
>     * the health Lara lost or gained
> 
> 
> between starting a particular level and dying and restarting it?

-   the pickups Lara collected
    - Pickups such as Lost Valley cogs you've picked up get removed.
-   the guns Lara found
     - Correctly set to the ones Lara started the level with.
-   the secrets Lara found
    - Secrets, time, and pickups correctly reset to 0.  
-   the ammo Lara wasted
    - Correctly set to the amount Lara started the level with.
-   the medpacks Lara consumed
    - Correctly set to the amount Lara started the level with.
-   the health Lara lost or gained
    - Didn't realize this was an experimental feature. This one rolls over dead Lara's HP. So you restart the level and instantly die haha. So this will need fixed.

So, I think just the health will need fixing if `disable_healing_between_levels` is true.

**Comment by rr-** [comment]
Oh wow, it's way better than I expected. How does this even work? :D

**Comment by walkawayy** [comment]
Pushed a WIP commit to add support for disable_healing_between_levels. There are some major flaws I noticed that I outlined in the commit.

**Comment by walkawayy** [comment]
> 
> 
> Oh wow, it's way better than I expected. How does this even work? :D Is SaveGame not being written to and holding the data from the last load? What about the following scenario:
> 
>     1. Save the game
> 
>     2. Collect pick up X
> 
>     3. Go to the next level
> 
>     4. Collect another pick up Y
> 
>     5. Die
> 
>     6. Restart the level
> 
> 
> Does Lara keep X?

1. Saved the game in Lost Valley.
2. Picked up a Lost Valley cog.
3. Used the level skip cheat.
4. There is no cog in Lara's inventory at the start of Tomb of Qualopec.
5. Collected the Scion.
6. Died to falling rubble.
7. Restarted the level.
8. There is no cog or Scion in Lara's inventory.
9. Loaded the original Lost Valley save and everything was normal.

**Comment by rr-** [comment]
I meant a pickup such as a medpack or ammo, that gets carried over between level restarts.

File: src/global/vars.h
Comment (rr-) [inline src/global/vars.h:151]: I don't think we need two variables to signify Lara's health – just one should've been enough?

File: src/game/option.c
Comment (rr-) [inline src/game/option.c:573]: This is sketchy, can we simplify it?

File: src/game/option.c
Comment (rr-) [inline src/game/option.c:533]: What does it do?

File: src/game/option.c
Comment (rr-) [inline src/game/option.c:502]: Missing braces

File: src/game/game.c
Comment (rr-) [inline src/game/game.c:58]: I'd reword it to `return GF_START_GAME | (InvMode == INV_DEATH_MODE ? CurrentLevel : GF.first_level_num);`

File: src/game/control.c
Comment (rr-) [inline src/game/control.c:250]: We probably should remove these before going through with the PR

**Review by rr- — COMMENTED** [review]
Let's also wait for a review from @thatguyfrombb

File: src/global/vars.h
Comment (walkawayy) [reply src/global/vars.h:151]: Well `StoredLaraHealth` gets updated every loop. So what was happening was, when Lara died, she would restart the level and instantly die on respawn since she had no health. Also, if  a restart level was ever implemented that could happen at any time, she would restart with her current health, not the health she entered the level with.

File: src/game/option.c
Comment (walkawayy) [reply src/game/option.c:573]: This was taking from Tomb2Main. Possibly? Why is it sketchy?

**Comment by walkawayy** [comment]
> 
> 
> I meant a pickup such as a medpack or ammo, that gets carried over between level restarts.

Ok double checked with pickups and it also works fine.
1. Saved the game in Lost Valley.
2. Picked up a large medipack.
3. Finished the level.
4. The large medipack is in Lara's inventory at the start of Tomb of Qualopec.
5. Collected a small medipack. Lara now has a small and large medipack in her inventory.
6. Died.
7. Restarted the Tomb of Qualopec level.
8. Lara has the large medipack in her inventory but not the small medipack.

**Comment by walkawayy** [comment]
This solution is on hold until #212 is implemented with level completion health. That is needed to save level completion health in order to fully support `disable_healing_between_levels`.

**Comment by walkawayy** [comment]
@rr- I have some questions on finishing this feature now that the new save game system is here. Currently it works when you restart the current level giving all the correct level starting items. However, I'm not sure how you'd like to make it work with the health carryover between level option. Also, there is a question on how to best handle this with legacy saves.

**Comment by rr-** [comment]
Most of the work is done in #409, I'm not sure if there should be anything left if you merge that TBH. With legacy saves we don't care, let's just make sure they restore full health.

**Comment by walkawayy** [comment]
Ya good point. I think my questions apply for when there's a Level Select feature.

**Comment by rr-** [comment]
For Level Select, we'll probably have to move inventory in its entirety to START INFO, as currently it's scattered between `inventory` and `levels`. Maybe it's best to do it now before we still haven't released, to not bother with backwards compatibility so early? As a separate ticket, of course.

**Comment by rr-** [comment]
Actually, we don't need to move the inventory. The only "global" inventory are key items which are strictly related to the current level. It doesn't make sense to carry over leadbars to the next level in the STARTINFO. As for the medpacks, guns, and ammo, they're all already in STARTINFO so we're almost good to go with the Level Select, maybe apart from the grabbed secrets/pickups/kills, which will be moved in #336 anyway.

File: src/game/control.c
Comment (walkawayy) [inline src/game/control.c:220]: Is this a relic of the old method? The one thing I wasn't sure about. The testing I did works, but maybe I should remove this and test again.

**Review by walkawayy — COMMENTED** [review]
This feature is ready now that the new savegame format is done.

File: src/game/option_passport.c
Comment (rr-) [inline src/game/option_passport.c:217]: What's this about?

File: src/global/vars.c
Comment (rr-) [inline src/global/vars.c:66]: Thanks for removing this.

**Review by rr- — APPROVED** [review]
The passport code is horrible. Thanks. Looks good, but will need some heavy testing. I'll try to get to it tomorrow :)

(Can we get rid of the passport? And make the player choose what save to load via json files)

File: src/game/option_passport.c
Comment (walkawayy) [reply src/game/option_passport.c:217]: If there's no saved games, you can't turn the page left from restart level to the load game page.

**Comment by rr-** [comment]
Still haven't tested but in the meantime let's add an entry in the changelog and the readme file :)

**Comment by rr-** [comment]
Tested and it's working fine. All that's missing is the changelog + readme entry

**Review by rr- — APPROVED** [review]
Let's resolve conflicts

**Pull request merged**
